src/watchdog.c

```diff
@@ -129,19 +145,33 @@ IMPLEMENT_MOCKABLE_FUNCTION(, void, watchdog_reset, WATCHDOG_HANDLE, watchdog)
     }
     else
     {
-        /*Codes_SRS_WATCHDOG_42_032: [ watchdog_reset shall set the state of the watchdog to STOP. ]*/
-        (void)interlocked_exchange(&watchdog->state, WATCHDOG_STOP);
-
         LogVerbose("Stopping watchdog for reset %p", watchdog);
 
-        /*Codes_SRS_WATCHDOG_42_033: [ watchdog_reset shall cancel the current timer by calling threadpool_timer_cancel. ]*/
-        threadpool_timer_cancel(watchdog->timer);
+        /*Codes_SRS_WATCHDOG_45_011: [ watchdog_reset shall call sm_close_begin. ]*/
+        if (sm_close_begin(watchdog->state) != SM_EXEC_GRANTED)
+        {
+            LogError("Watchdog could not be closed for reset");
+        }
+        else /*Codes_SRS_WATCHDOG_45_018: [ If sm_close_begin returns SM_EXEC_GRANTED, ]*/
+        {
+            /*Codes_SRS_WATCHDOG_42_033: [ watchdog_reset shall cancel the current timer by calling threadpool_timer_cancel. ]*/
+            threadpool_timer_cancel(watchdog->timer);
+            /*Codes_SRS_WATCHDOG_45_012: [ watchdog_reset shall call sm_close_end. ]*/
+            sm_close_end(watchdog->state);
+        }
 
         LogVerbose("Restarting watchdog %p", watchdog);
-
-        /*Codes_SRS_WATCHDOG_42_034: [ watchdog_reset shall set the state of the watchdog to RUNNING. ]*/
-        (void)interlocked_exchange(&watchdog->state, WATCHDOG_RUNNING);
-
+        /*Codes_SRS_WATCHDOG_45_013: [ watchdog_reset shall call sm_open_begin. ]*/
+        if (sm_open_begin(watchdog->state) != SM_EXEC_GRANTED)
+        {
+            LogError("Watchdog could not be reopened for reset");
+        }
+        else
+        {
+            /*Codes_SRS_WATCHDOG_45_014: [ watchdog_reset shall call sm_open_end if sm_open_begin succeeds. ]*/
+            sm_open_end(watchdog->state, true);
+        }
+        
         /*Codes_SRS_WATCHDOG_42_035: [ watchdog_reset shall restart the timer by calling threadpool_timer_restart with the original timeout_ms from the call to start. ]*/
         (void)threadpool_timer_restart(watchdog->timer, watchdog->timeout_ms, 0);

```

Comment:
>threadpool_timer_restart [](http://example.com/codeflow?start=14&length=24)

If we were not able to do sm_open_begin, then we must be in a weird state...not sure we should bother restarting the timer in that case...

Reply:
Good point - done.